plutus-core/executables/pir/Main.hs

```diff
@@ -1,106 +1,288 @@
-loadPir = loadASTfromFlat Named
-
-compile :: COpts -> PirProg () -> Either PIRError PLCTerm
-compile opts (PIR.Program _ pirT) = do
+
+type PirError a = PIR.Error PLC.DefaultUni PLC.DefaultFun a
+type UnitProvenance = PIR.Provenance ()
+
+---------------- Types for command line options ----------------
+
+-- | A specialised format type for PIR. We don't support deBruijn or named deBruijn for PIR.
+data PirFormat = TextualPir | FlatNamed
+instance Show PirFormat
+    where show = \case { TextualPir  -> "textual"; FlatNamed -> "flat-named" }
+
+data PirOptimiseOptions = PirOptimiseOptions Input PirFormat Output PirFormat PrintMode
+
+data AnalyseOptions = AnalyseOptions Input PirFormat Output -- Input is a program, output is text
+
+data Language = PLC | UPLC
+instance Show Language
+    where show = \case { PLC  -> "plc"; UPLC -> "uplc" }
+
+-- | Compilation options: target language, whether to optimise or not, input and output streams and types
+data CompileOptions =
+    CompileOptions Language
+                   Bool   -- Optimise or not?
+                   Bool   -- True -> just report if compilation was successful; False -> write output
+                   Input
+                   PirFormat
+                   Output
+                   Format
+                   PrintMode
+
+data Command = Analyse  AnalyseOptions
+             | Compile  CompileOptions
+             | Convert  ConvertOptions
+             | Optimise PirOptimiseOptions
+             | Print    PrintOptions
+
+
+---------------- Option parsers ----------------
+
+-- | Invert a switch: return False if the switch is supplied on the command
+-- line, True otherwise.  This is used for command-line options to turn things
+-- off.
+switch' :: Mod FlagFields Bool -> Parser Bool
+switch' = fmap not . switch
+
+pirFormatHelp :: String
+pirFormatHelp =
+  "textual or flat-named (names)"
+
+pirFormatReader :: String -> Maybe PirFormat
+pirFormatReader =
+    \case
+         "textual"    -> Just TextualPir
+         "flat-named" -> Just FlatNamed
+         "flat"       -> Just FlatNamed
+         _            -> Nothing
+
+pPirInputFormat :: Parser PirFormat
+pPirInputFormat = option (maybeReader pirFormatReader)
+  (  long "if"
+  <> long "input-format"
+  <> metavar "PIR-FORMAT"
+  <> value TextualPir
+  <> showDefault
+  <> help ("Input format: " ++ pirFormatHelp))
+
+pPirOutputFormat :: Parser PirFormat
+pPirOutputFormat = option (maybeReader pirFormatReader)
+  (  long "of"
+  <> long "output-format"
+  <> metavar "PIR-FORMAT"
+  <> value TextualPir
+  <> showDefault
+  <> help ("Output format: " ++ pirFormatHelp))
+
+pPirOptimiseOptions :: Parser PirOptimiseOptions
+pPirOptimiseOptions = PirOptimiseOptions <$> input <*> pPirInputFormat <*> output <*> pPirOutputFormat <*> printmode
+
+pAnalyseOptions :: Parser AnalyseOptions
+pAnalyseOptions = AnalyseOptions <$> input <*> pPirInputFormat <*> output
+
+languageReader :: String -> Maybe Language
```

Comment:
I feel like all `Language` related functions should go in `Common.hs`? It may be useful for other executables too later and it's not specific to PIR?